benchmarks/cpp/nvfuser/normalization.cpp

```diff
@@ -0,0 +1,638 @@
+  auto norm = mul(x_mean_sub, rvar);
+
+  auto weight_bcast = broadcast(weight, broadcast_mask);
+  auto bias_bcast = broadcast(bias, broadcast_mask);
+  auto norm_gamma = mul(norm, weight_bcast);
+  auto norm_gamma_bias = add(norm_gamma, bias_bcast);
+  return norm_gamma_bias;
+}
+
+static TensorView* myLayerNorm(Fusion* fusion,
+                        TensorView* input,
+                        const int kNumberOfDims,
+                        std::vector<int64_t>& norm_shape) {
+  FusionGuard fg(fusion);
+
+  const float kEps = 1e-5;
+  std::vector<int> reduction_axes(norm_shape.size());
+  std::vector<bool> broadcast_mask(input->nDims(), false);
+  torch::jit::fuser::cuda::Val* num_features = nullptr;
+  for (int idx = 0; idx < norm_shape.size(); ++idx) {
+    const int axis = input->nDims() - 1 - idx;
+    reduction_axes[idx] = axis;
+    broadcast_mask[axis] = true;
+    num_features = (num_features == nullptr)
+        ? input->domain()->domain()[axis]->extent()
+        : mul(num_features, input->domain()->domain()[axis]->extent());
+  }
+
+  // Reduction
+  auto x_sum = sum(input, reduction_axes);
+  // Broadcast
+  auto x_sum_bcast = broadcast(x_sum, broadcast_mask);
+  // Point-wise
+  auto x_mean = div(x_sum_bcast, num_features);
+  auto x_mean_sub = sub(input, x_mean);
+
+  auto x_mean_sub_pow = mul(x_mean_sub, x_mean_sub);
+  // Reduction
+  auto var_sum = sum(x_mean_sub_pow, reduction_axes);
+  // Broadcast
+  auto var_sum_bcast = broadcast(var_sum, broadcast_mask);
+  // Point-wise
+  auto var = div(var_sum_bcast, num_features);
+  auto var_eps = add(var, new Float(kEps));
+  auto rvar = unaryOp(UnaryOpType::Rsqrt, var_eps);
+  auto output = mul(x_mean_sub, rvar);
+  return output;
+}
+
+
+//------------------------------------------------------------------------------
+
+static void MagicScheduler_Softmax(benchmark::State& benchmark_state) {
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  std::vector<int64_t> input_shape{benchmark_state.range(1), benchmark_state.range(0)};
+  const int kReductionAxis = benchmark_state.range(2);
+
+  // setup fusion
+  auto input = TensorViewBuilder().ndims(input_shape.size()).dtype(DataType::Float).build();
+  fusion.addInput(input);
+  auto output = mySoftmax(&fusion, input, input_shape.size(), kReductionAxis);
+  fusion.addOutput(output);
+
+  std::vector<TensorView*> reduction_tensors;
+  std::vector<TensorView*> other_tensors;
+  analyzeFusion(&fusion, reduction_tensors, other_tensors);
+
+  // inputs
+  at::manual_seed(0);
+  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
+  at::Tensor at_x = at::randn(input_shape, options);
+  std::vector<c10::IValue> inputs({at_x});
+
+  // outputs
+  std::vector<at::Tensor> outputs;
+
+  auto reduction_params =
+      getNormalizationHeuristics(&fusion, inputs, reduction_tensors);
+  TORCH_CHECK(reduction_params, "Reduction schedule was not generated!");
+
+  scheduleNormalization(
```

Comment:
would it make sense to benchmark `scheduleNormalization` by itself (as a separate benchmark case) ?